src/Mods.jl

```diff
@@ -49,224 +49,130 @@ julia> value(a)
-
-conj(x::Mod) = x   # so matrix transpose will work
-
-function hash(x::AbstractMod, h::UInt64= UInt64(0))
+function hash(x::Mod, h::UInt64= UInt64(0))
     v = value(x)
     m = modulus(x)
     return hash(v,hash(m,h))
 end
 
 # Test for equality
-isequal(x::Mod, y::Mod) = value(x)==value(y) && modulus(x)==modulus(y)
-
-==(x::Mod,y::Mod) = isequal(x,y)
-
-
-"""
-`modcheck(x::Mod,y::Mod)` checks if `x` and `y` have the same modulus.
-If so, return than modulus. If not, throw an error.
-"""
-function modcheck(x::Mod, y::Mod)::Int
-    mx = modulus(x)
-    my = modulus(y)
-    if mx != my
-        error("Cannot operate on two Mod objects with different moduli")
-    end
-    return mx
-end
+==(x::Mod{N,T1}, y::Mod{M,T2}) where {M,N,T1,T2} = false
+==(x::Mod{N,T1}, y::Mod{N,T2}) where {N,T1,T2} = value(x)==value(y)
 
 # Easy arithmetic
-function +(x::Mod, y::Mod)::Mod
-    m = modcheck(x,y)
-    s,flag = Base.add_with_overflow(x.val,y.val)
+@inline function +(x::Mod{N,T}, y::Mod{N,T}) where {N,T}
+    s, flag = Base.add_with_overflow(x.val,y.val)
     if !flag
-        return Mod(x.val+y.val, m)
+        return Mod{N,T}(s)
     end
-    s = widen(x.val) + widen(y.val)    # add with added precision
-    s = Int(mod(s,m))                  # reduce by modulus
-    return Mod(s,m)
+    t = widen(x.val) + widen(y.val)    # add with added precision
+    return Mod{N,T}(unsafe_mod(t,N))
 end
 
 
-function -(x::Mod{M}) where M
-    return Mod(-x.val, M)
+function -(x::Mod{M,T}) where {M,T<:Signed}
+    return Mod{M,T}(-x.val)  # Note: might break for UInt
+end
+
+function -(x::Mod{M,T}) where {M,T<:Unsigned}
+    return Mod{M,T}(M-value(x))
 end
 
 -(x::Mod,y::Mod) = x + (-y)
 
+unsafe_mod(x, y) = x - (x ÷ y) * y  # does not check `y` being positive
 
-function *(x::Mod, y::Mod)
-    m = modcheck(x,y)
-    p,flag = Base.mul_with_overflow(x.val,y.val)
+@inline function *(x::Mod{N,T}, y::Mod{N,T}) where {N,T}
+    p, flag = Base.mul_with_overflow(x.val,y.val)
     if !flag
-        return Mod(x.val*y.val, m)
+        return Mod{N,T}(p)
+    else
+        q = widemul(x.val, y.val)         # multipy with added precision
+        return Mod{N,T}(unsafe_mod(q,N)) # return with proper type
     end
-    p = widemul(x.val, y.val)         # multipy with added precision
-    p = Int(mod(p,m))                  # reduce by the modulus
-    return Mod(p,m) # return with proper type
 end
 
 # Division stuff
 """
 `is_invertible(x::Mod)` determines if `x` is invertible.
 """
 function is_invertible(x::Mod{M})::Bool where M
-    return gcd(value(x),M) == 1
+    return gcd(x.val,M) == 1
 end
 
 
 """
 `inv(x::Mod)` gives the multiplicative inverse of `x`.
 This may be abbreviated by `x'`.
 """
-function inv(x::Mod{M}) where M
-    (g, v, ignore) = gcdx(x.val, M)
+@inline function inv(x::Mod{M,T}) where {M,T}
+    Mod{M,T}(_invmod(x.val, M))
+end
+_invmod(x::Unsigned, m::Unsigned) = invmod(x, m)
+@inline function _invmod(x::Signed, m::Signed)
```

Comment:
Good point, `gcdx` is faster, but not applicable for UInt. Using `invmod` in both cases will cause regression in performance.

Reply:
I see. `invmod` has to check special cases wit negative arguments and performs a `mod`, which are not necessary in this use case. Maybe write a comment to avoid this question in the future,

Reply:
Sounds good, added some comments.